stdlib/format.ml

```diff
@@ -1026,18 +1026,45 @@ module DLS = Domain.DLS
 let stdbuf_key = DLS.new_key pp_make_buffer
 let _ = DLS.set stdbuf_key stdbuf
 
+let str_formatter_key = DLS.new_key (fun () ->
+  formatter_of_buffer (DLS.get stdbuf_key))
+let _ = DLS.set str_formatter_key str_formatter
+
+let buffered_out_string key str ofs len =
+  Buffer.add_substring (Domain.DLS.get key) str ofs len
+
+let buffered_out_flush oc key () =
+  let buf = Domain.DLS.get key in
+  let len = Buffer.length buf in
+  let str = Buffer.contents buf in
+  output_substring oc str 0 len ;
+  Buffer.clear buf
+
+let std_buf_key = Domain.DLS.new_key (fun () -> Buffer.create 80)
+let err_buf_key = Domain.DLS.new_key (fun () -> Buffer.create 80)
+
 let std_formatter_key = DLS.new_key (fun () ->
-  formatter_of_out_channel Stdlib.stdout)
+  let ppf =
+    pp_make_formatter (buffered_out_string std_buf_key)
+      (buffered_out_flush Stdlib.stdout std_buf_key) ignore ignore ignore
+  in
+  ppf.pp_out_newline <- display_newline ppf;
+  ppf.pp_out_spaces <- display_blanks ppf;
+  ppf.pp_out_indent <- display_indent ppf;
+  ppf)
```

Comment:
I shall update the documentation to say that 

```
the output is interleaved at points where the formatter is flushed
```

I like the current formatter for the reason that we can do multi-line printing from a domain and not have it interleaved with output from other domains. I don't know whether this is a reasonable usecase, but I may wish to the produce log from multiple domains as a well-formed markdown file:

```md
* Domain 0 did Event e
  + Variable a was xxx
  + Variable b was yyy
```

Assuming that a flush is done after the whole event is emitted, the current version will produce unsurprising behaviours.  The same would be true for emitting well-formatted json, where each event is an item in an array.

Reply:
I agree, and think that such use cases are legit. My only concern is that there is a potential for code that produces a lot of output before flushing to consume significantly more memory when run with such a buffered formatter. This could be memory consumption with asymptotic complexity that is linear in the number of execution steps. It can be avoided, but I'm not sure how it fits with the goal of how unchanged sequential code runs on multicore vs the sequential runtime. If there is a goal for existing code to not take asymptotically more memory on multicore, then the risk could be reduced by weakening to interleaving at newlines. I don't know the right trade-off.